pool.go

```diff
@@ -0,0 +1,273 @@
+}
+
+func (p *Pool) handleError(err error) {
+	if p.Opts.ErrorHandler == nil {
+		return
+	}
+
+	go p.Opts.ErrorHandler(err)
+}
+
+// Connect creates poll of connections by calling Factory method and connect them all
+func (p *Pool) Connect() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.isClosed {
+		return errors.New("pool is closed")
+	}
+
+	// build connections
+	for _, addr := range p.Addrs {
+		conn, err := p.Factory(addr)
+		if err != nil {
+			return fmt.Errorf("creating connection for %s: %w", addr, err)
+		}
+
+		// set own handler when connection is closed
+		conn.SetOptions(ConnectionClosedHandler(p.handleClosedConnection))
+
+		err = conn.Connect()
+		if err != nil {
+			p.handleError(fmt.Errorf("connecting to %s: %w", conn.addr, err))
+			p.wg.Add(1)
+			go p.recreateConnection(conn)
+			continue
+		}
+
+		p.connections = append(p.connections, conn)
+	}
+
+	if len(p.connections) < p.Opts.MinConnections {
+		return fmt.Errorf("minimum %d connections is required, established: %d", p.Opts.MinConnections, len(p.connections))
+	}
+
+	return nil
+}
+
+// Connections returns copy of all connections from the pool
+func (p *Pool) Connections() []*Connection {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// create new slice, as p.connections can be changed
+	var conns []*Connection
+	conns = append(conns, p.connections...)
+
+	return conns
+}
+
+// FilterFunc is a function to filter connections
+type FilterFunc func(*Connection) bool
+
+// Get returns filtered connection from the pool
+func (p *Pool) Get() (*Connection, error) {
+	p.mu.Lock()
+	if p.isClosed {
+		p.mu.Unlock()
+		return nil, errors.New("pool is closed")
+	}
+	p.mu.Unlock()
+
+	// filtered connections
+	conns := p.filteredConnections()
+
+	if len(conns) == 0 {
+		return nil, errors.New("no (filtered) connections in the pool")
+	}
+
+	n := atomic.AddUint32(&p.connIndex, 1)
+
+	return conns[(int(n)-1)%len(conns)], nil
+}
+
+// when connection is closed, remove it from the pool of connections and start
+// goroutine to create new connection for the same address
+func (p *Pool) handleClosedConnection(closedConn *Connection) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	var connIndex = -1
+	for i, conn := range p.connections {
+		if conn == closedConn {
+			connIndex = i
+			break
+		}
+	}
+
+	// somehow we didn't find closed connection in the pool
+	if connIndex < 0 {
+		p.handleError(errors.New("closed connection was not found in the pool"))
+		return
+	}
+
+	p.connections[connIndex] = p.connections[len(p.connections)-1] // Copy last element to index connIndex.
+	p.connections[len(p.connections)-1] = nil                      // Erase last element
+	p.connections = p.connections[:len(p.connections)-1]           // Truncate slice.
+
+	// if pool was closed, don't start recreate goroutine
+	// should we return earlier and keep connection in the pool as it will be closed anyway?
+	if p.isClosed {
+		return
+	}
+
+	// initiate goroutine to reconnect to closedConn.Addr
+	p.wg.Add(1)
```

Comment:
It seems that this is required in all cases before you call 
go p.recreateConnection(closedConn)

To me that indicates it should be folded into it

Reply:
wg.Add should be called outside of the gorutine. Here is an answer why: https://stackoverflow.com/a/65213822

Reply:
Can you not wrap it in a package private method that first does the add and then spawns a goroutine?

Reply:
Not sure I understand the question.